dq.go

```diff
@@ -0,0 +1,212 @@
+			apis := v.Meta.(map[string]interface{})["Apis"]
+
+			// I hate this
+			var exp int64
+			switch v.Meta.(map[string]interface{})["QuotaRenewal"].(type) {
+			case int64:
+				exp = v.Meta.(map[string]interface{})["QuotaRenewal"].(int64)
+			case float64:
+				exp = int64(v.Meta.(map[string]interface{})["QuotaRenewal"].(float64))
+			}
+
+			expT := time.Unix(exp, 0)
+			for _, aid := range apis.([]interface{}) {
+				apiID := aid.(string)
+
+				// This will grab the session handler
+				spec := GetSpecForApi(apiID)
+				if spec == nil {
+					log.Warning("Can't find back-end for this API, skippingc")
+					break
+				}
+
+				// Have we processed on this handler before (many APIs may use the same handler)?
+				_, processedOnSH := processedSpecs[spec.SessionManager]
+				if !processedOnSH {
+					// This handler hasn't been used yet for the API
+					// Get the session data
+					s, f := spec.SessionManager.GetSessionDetail(k)
+
+					// If it was found, lets process it for this handler
+					if f {
+						skip := false
+						if time.Now().After(expT) {
+							QuotaHandler.TagDelete(k)
+							skip = true
+						}
+
+						if s.IsExpired() {
+							// Remove expired data too
+							QuotaHandler.TagDelete(k)
+							skip = true
+						}
+
+						if !skip {
+							qr := int64(v.Max - v.Counter.Count())
+							if qr < 0 {
+								qr = 0
+							}
+
+							// Only write on count difference
+							if qr != s.QuotaRemaining {
+								s.QuotaRemaining = qr
+								spec.SessionManager.UpdateSession(k, s, GetLifetime(&dummyAPISpec, &s))
+								// Since we've written the token, we don't need to re-do it at the end of the middleware chain
+								s.SetFirstSeenHash()
+								// We've performed a write on this SH now, lets tag that so we don't do it again
+								processedSpecs[spec.SessionManager] = struct{}{}
+							}
+						}
+
+					} else {
+						// No longer in session store, delete
+						QuotaHandler.TagDelete(k)
+					}
+				}
+			}
+		})
+	}
+
+	return nil
+
+}
+
+func StartDQ(statusFunc GetLeaderStatusFunc) {
+	log.WithFields(logrus.Fields{
+		"prefix": "DQuota",
+	}).Info("Using Distributed Quota")
+	p := strconv.Itoa(config.Storage.Port)
+	cs := fmt.Sprintf("redis://%v:%v", config.Storage.Host, p)
+	c1, _ := client.NewClient(cs, encoding.JSON)
+
+	QuotaHandler = dq.NewDQ(DQFlusher, DQErrorHandler, NodeID)
+	QuotaHandler.BroadcastWith(c1, time.Millisecond*100, getDQTopic())
+
+	// We always need a leader because otherwise we can;t persist data
+	QuotaHandler.SetLeader(statusFunc())
+
+	QuotaHandler.FlushInterval = time.Second * 3
+	if config.DistributedQuotaFlushIntervalInMS != 0 {
+		QuotaHandler.FlushInterval = time.Millisecond * time.Duration(config.DistributedQuotaFlushIntervalInMS)
+	}
+
+	DQFlusherPool.Open()
+
+	if err := QuotaHandler.Start(); err != nil {
+		log.Fatal(err)
+	}
+
+	// Give us time to catch up
+	time.Sleep(time.Millisecond * 100)
```

Comment:
IIRC it makes sure that the flusher and the CDRT notifier has time to initialise and do it's thing. There are two systems working in tandem here, one is a batch writer and the other is a distributed in-memory merge, the two are eventually consistent.

This function is only called at startup, so this pause is ok.